golang/cosmos/x/vtransfer/ibc_middleware.go

```diff
@@ -0,0 +1,151 @@
+package vtransfer
+
+import (
+	"github.com/Agoric/agoric-sdk/golang/cosmos/x/vtransfer/keeper"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	capabilitytypes "github.com/cosmos/cosmos-sdk/x/capability/types"
+	clienttypes "github.com/cosmos/ibc-go/v6/modules/core/02-client/types"
+	channeltypes "github.com/cosmos/ibc-go/v6/modules/core/04-channel/types"
+	porttypes "github.com/cosmos/ibc-go/v6/modules/core/05-port/types"
+	"github.com/cosmos/ibc-go/v6/modules/core/exported"
+)
+
+var _ porttypes.Middleware = (*IBCMiddleware)(nil)
```

Comment:
> I am very confused about delegation to `im.vtransferKeeper` vs. `im.app` vs. both.

I wonder if this would be better handled by forwarding *everything* to the other file, which can better explain itself—i.e., that every method of this file's IBCMiddleware looks like so:
```go
// $FUNCTION_NAME implements the $INTERFACE_NAME interface
func (wrapper IBCMiddleware) $FUNCTION_NAME(
	ctx sdk.Context,
	…,
) (…) {
	return wrapper.vtransferKeeper.$FUNCTION_NAME(ctx, wrapper.innerIBCModule, …)
}
```

But digging further, it seems to me me that using the vtransfer "keeper" in this way is an abuse of that concept, "[_whose role is to manage access to the subset of the **state** defined by [its module]_](https://docs.cosmos.network/v0.46/building-modules/keeper.html#keepers)" (emphasis mine). In other words, keepers exist to wrap [IAVL] _stores_ rather than arbitrary module logic AFAICT. So really, we can omit that indirection altogether and put the logic _here_, using the keeper only for `Has`/`Set`/`Delete`, e.g.:
```go
// OnAcknowledgementPacket checks if the packet contains a contract
// invocation, and if so, sends it to the vIBC module.
// It implements the IBCModule interface.
func (wrapper IBCMiddleware) OnAcknowledgementPacket(
	ctx sdk.Context,
	packet channeltypes.Packet,
	acknowledgement []byte,
	relayer sdk.AccAddress,
) error {
	// Always trigger the wrapped code.
	innerErr := wrapper.innerIBCModule.OnAcknowledgementPacket(ctx, packet, acknowledgement, relayer)

	// Forward to vIBC, even if the wrapped code already returned an error.
	// XXX BUT WHY?
	sender, _, err := wrapper.vibcKeeper.parseTransfer(ctx, packet)
	if err == nil && sender != "" {
		err = k.vibcKeeper.TriggerOnAcknowledgementPacket(ctx, sender, packet, acknowledgement, relayer)
	}

	// innerErr trumps later errors.
	if innerErr != nil {
		return innerErr
	}
	return err
}
```